File: cpp/sparse.cpp
#include <iomanip>
#include "state.hpp"

int main(int argc, char** argv) {
    std::string path = argv[1];
    Hamiltonian H = Hamiltonian::from_file(path);

    FullState state(H.paulis_list());
    auto [stab, energy_gs] = state.evolve();
    std::cout << "ground state stabilizers" << std::endl;
    std::cout << stab.range(0, H.n).to_string(true) << std::endl;
    std::cout << std::setprecision(16) << "ground state energy " << energy_gs << std::endl;
    std::cout << "terms signs" << std::endl;
    for (auto [P, w]: H.original_paulis) {
        std::cout << stab.energy(P) << " ";
    }
    std::cout << std::endl;
}

File: py/sparse.py
import sys
from stab import StabilizerGroup
from state import Hamiltonian, FullStateEnergies

H = Hamiltonian.from_file(sys.argv[1])
print(*H.original_paulis, sep='\n')

# use sparse (and nonlocal) algorithm
fullstateE = FullStateEnergies(H.original_paulis)
#fullstateE.evolve()
full_Ps, full_energy_gs = fullstateE.evolve()
full_stab = StabilizerGroup(full_Ps)
print('gs energy', full_energy_gs)
# stabilizer group S
print(full_stab.Ps)
